zebrad/src/commands/start/sync.rs

```diff
@@ -0,0 +1,204 @@
+use color_eyre::Report;
+use eyre::eyre;
+use futures::stream::{FuturesUnordered, StreamExt};
+use std::{collections::HashSet, iter};
+use tower::{Service, ServiceExt};
+use tracing_futures::Instrument;
+use zebra_chain::{block::BlockHeaderHash, types::BlockHeight};
+
+pub struct Syncer<ZN, ZS>
+where
+    ZN: Service<zebra_network::Request>,
+{
+    pub peer_set: ZN,
+    pub state: ZS,
+    pub tip_requests: FuturesUnordered<ZN::Future>,
+    pub block_requests: FuturesUnordered<ZN::Future>,
+    pub block_locator: Vec<BlockHeaderHash>,
+    pub downloading: HashSet<BlockHeaderHash>,
+    pub downloaded: HashSet<BlockHeaderHash>,
+    pub fanout: NumReq,
+}
+
+impl<ZN, ZS> Syncer<ZN, ZS>
+where
+    ZN: Service<zebra_network::Request, Response = zebra_network::Response, Error = Error>
+        + Send
+        + Clone
+        + 'static,
+    ZN::Future: Send,
+    ZS: Service<zebra_state::Request, Response = zebra_state::Response, Error = Error>
+        + Send
+        + Clone
+        + 'static,
+    ZS::Future: Send,
+{
+    pub async fn run(&mut self) -> Result<(), Report> {
+        loop {
+            if self.tip_requests.is_empty() {
+                info!("populating prospective tips list");
+                self.obtain_tips(vec![super::GENESIS]).await?;
```

Comment:
This block locator will work only in the first iteration of the loop, but not afterwards.  The block locator needs to be set to
```
[tip, tip-1, tip-2, ..., tip-9, tip-20, tip-40, tip-80, tip-160 ]
```
or similar so that we ask for new tips starting from our current state, not always from genesis.  When we're doing initial block download, the state should consist just of the genesis block, so this would be `[genesis]` since that's the first chain tip.

Reply:
In fact, constructing the block locator might be cleaner to do in `obtain_tips`, since we need to do it every time that function is called.

Reply:
Do we need to specify different genesis block hashes, based on the network we're using?
(The main or test networks?)

Reply:
We don't have to do anything network-specific if we construct the block locator by querying the state in `obtain_tips`.

Reply:
Ah, that's not true, we need the genesis block in order to do rejections.

Reply:
I suggest that we reject both genesis blocks, in case a peer or seeder is misconfigured.

Otherwise, we'll build both chains, which is *not good*.

Reply:
I updated #504 to reject both genesis blocks.